code/modules/antagonists/eldritch_cult/knowledge/rust_lore.dm

```diff
@@ -0,0 +1,222 @@
+	gain_text = "Champion of rust. Corruptor of steel. Fear the dark for Rustbringer has come!"
+	cost = 3
+	required_atoms = list(/mob/living/carbon/human)
+	route = "Rust"
+	var/finished = FALSE
+	///keeps track of previous brute mod
+	var/prev_brute_mod
+	///keeps track of previous burn mod
+	var/prev_burn_mod
+
+/datum/eldritch_knowledge/rust_final/recipe_snowflake_check(list/atoms, loc,list/selected_atoms)
+	if(finished)
+		return FALSE
+	var/counter = 0
+	for(var/mob/living/carbon/human/H in atoms)
+		selected_atoms |= H
+		counter++
+		if(counter == 3)
+			return TRUE
+	return FALSE
+
+/datum/eldritch_knowledge/rust_final/on_finished_recipe(mob/living/user, list/atoms, loc)
+	var/mob/living/carbon/human/H = user
+	prev_brute_mod = H.physiology.brute_mod
+	prev_burn_mod = H.physiology.burn_mod
+	finished = TRUE
+	priority_announce("$^@&#*$^@(#&$(@&#^$&#^@# Fear the decay, for Rustbringer [user.real_name] has come! $^@&#*$^@(#&$(@&#^$&#^@#","#$^@&#*$^@(#&$(@&#^$&#^@#", 'sound/ai/spanomalies.ogg')
+	new /datum/rust_spread(loc)
+
+	return ..()
+
+
+/datum/eldritch_knowledge/rust_final/on_life(mob/user)
+	. = ..()
+	if(!finished)
+		return
+	var/mob/living/carbon/human/H = user
+	var/turf/T = get_turf(user)
+	if(!istype(T,/turf/open/floor/plating/rust) || !isliving(user))
+		H.physiology.brute_mod = prev_brute_mod
+		H.physiology.burn_mod = prev_burn_mod
+		return
+
+	H.adjustBruteLoss(-3)
+	H.adjustFireLoss(-3, FALSE)
+	H.adjustToxLoss(-3, FALSE)
+	H.adjustOxyLoss(-1, FALSE)
+	H.adjustStaminaLoss(-10)
+	H.physiology.brute_mod = prev_brute_mod * 0.5
+	H.physiology.burn_mod = prev_burn_mod * 0.5
+
+
+/**
+  * #Rust spread datum
+  *
+  * Simple datum that automatically spreads rust around it
+  *
+  * Simple implementation of automatically growing entity
+  */
+/datum/rust_spread
+	var/list/edge_turfs = list()
+	var/list/turfs = list()
+	var/static/list/blacklisted_turfs = typecacheof(list(/turf/closed/wall/rust,/turf/closed/wall/r_wall/rust,/turf/open/space,/turf/open/lava,/turf/open/chasm,/turf/open/floor/plating/rust))
+	var/spread_per_tick = 5
+
+
+/datum/rust_spread/New(loc)
+	. = ..()
+	var/turf/T = get_turf(loc)
+	T.ChangeTurf(/turf/open/floor/plating/rust)
+	turfs += T
+	START_PROCESSING(SSprocessing,src)
+
+
+/datum/rust_spread/Destroy(force, ...)
+	STOP_PROCESSING(SSprocessing,src)
+	return ..()
+
+/datum/rust_spread/process()
+	compile_turfs()
+	var/turf/T
+	for(var/i in 0 to 6)
+		T = pick(edge_turfs)
+		T.rust_heretic_act()
+		turfs += get_turf(T)
+
+/**
+  * Compile turfs
+  *
+  * Recreates all edge_turfs as well as normal turfs.
+  */
+/datum/rust_spread/proc/compile_turfs()
+	edge_turfs = list()
+	for(var/X in turfs)
+		if(!istype(X,/turf/closed/wall/rust) && !istype(X,/turf/closed/wall/r_wall/rust) && !istype(X,/turf/open/floor/plating/rust))
+			turfs -=X
+			continue
+		for(var/turf/T in range(1,X))
+			if(is_type_in_typecache(T,blacklisted_turfs))
+				continue
+			edge_turfs += T
```

Comment:
Won't this select again previously selected turfs, repeating them on the list?

Reply:
blacklisted turfs has rusted walls and rusted floors unless i am mistaken

Reply:
We need to re-write this to be a bit more sane.
Try keeping a list of the edge turfs on the datum, and check it for length on the process. Pick from the existing list one at a time, `spread_per_tick` times, and recompile it with one more range once the list is empty. And so on.
Poke me if you need an algorithm to get range turfs without using `range()`.